### ISSUE TITLE
java.lang.NullPointerException - LineChartRenderer.drawValues(SourceFile:568)

### ISSUE THREAD

**Issue by eakteam** [PhilJay/MPAndroidChart#3714]
How to fix this line to solve NPE ???

THE LINE THAT CAUSES CRASH IN THE METHOD : 

```
Entry entry = dataSet.getEntryForIndex(j / 2 + mXBounds.min);

                    if (dataSet.isDrawValuesEnabled()) {
                        drawValue(c, dataSet.getValueFormatter(), entry.getY(), entry, i, x,
                                y - valOffset, dataSet.getValueTextColor(j / 2)); // THE CRASH HAPPEN SOMETIMES
                    }
```

THIS IS FULL METHOD : 

```
public void drawValues(Canvas c) {

        if (isDrawingValuesAllowed(mChart)) {

            List<ILineDataSet> dataSets = mChart.getLineData().getDataSets();

            for (int i = 0; i < dataSets.size(); i++) {

                ILineDataSet dataSet = dataSets.get(i);

                if (!shouldDrawValues(dataSet))
                    continue;

                // apply the text-styling defined by the DataSet
                applyValueTextStyle(dataSet);

                Transformer trans = mChart.getTransformer(dataSet.getAxisDependency());

                // make sure the values do not interfear with the circles
                int valOffset = (int) (dataSet.getCircleRadius() * 1.75f);

                if (!dataSet.isDrawCirclesEnabled())
                    valOffset = valOffset / 2;

                mXBounds.set(mChart, dataSet);

                float[] positions = trans.generateTransformedValuesLine(dataSet, mAnimator.getPhaseX(), mAnimator
                        .getPhaseY(), mXBounds.min, mXBounds.max);

                MPPointF iconsOffset = MPPointF.getInstance(dataSet.getIconsOffset());
                iconsOffset.x = Utils.convertDpToPixel(iconsOffset.x);
                iconsOffset.y = Utils.convertDpToPixel(iconsOffset.y);

                for (int j = 0; j < positions.length; j += 2) {

                    float x = positions[j];
                    float y = positions[j + 1];

                    if (!mViewPortHandler.isInBoundsRight(x))
                        break;

                    if (!mViewPortHandler.isInBoundsLeft(x) || !mViewPortHandler.isInBoundsY(y))
                        continue;

                    Entry entry = dataSet.getEntryForIndex(j / 2 + mXBounds.min);

                    if (dataSet.isDrawValuesEnabled()) {
                        drawValue(c, dataSet.getValueFormatter(), entry.getY(), entry, i, x,
                                y - valOffset, dataSet.getValueTextColor(j / 2));
                    }

                    if (entry.getIcon() != null && dataSet.isDrawIconsEnabled()) {

                        Drawable icon = entry.getIcon();

                        Utils.drawImage(
                                c,
                                icon,
                                (int)(x + iconsOffset.x),
                                (int)(y + iconsOffset.y),
                                icon.getIntrinsicWidth(),
                                icon.getIntrinsicHeight());
                    }
                }

                MPPointF.recycleInstance(iconsOffset);
            }
        }
    }
```

**Comment by diareuse** [comment]
Bug inside the library. Duplicate to #3696

**Comment by BTrottAtWork** [comment]
FYI https://github.com/PhilJay/MPAndroidChart/issues/3681

Entry needs to implement .equals.

**Comment by eakteam** [comment]
When i add the method method to the Entry Class how to reference it from LineChartRenderValues.drawValues ?

**Comment by eakteam** [comment]
Hello i still have the exception after overriding the methods. this is the Entry class : 

    ```
public boolean equalTo(Entry e) {

        if (e == null)
            return false;

        if (e.getData() != this.getData())
            return false;

        if (Math.abs(e.x - this.x) > Utils.FLOAT_EPSILON)
            return false;

        if (Math.abs(e.getY() - this.getY()) > Utils.FLOAT_EPSILON)
            return false;

        return true;
    }

    // add the following to the Entry class
    @Override
    public boolean equals(Object obj) {
        return obj != null && obj instanceof Entry && this.equalTo((Entry) obj);
    }

```

**Comment by eakteam** [comment]
The issue is still active, i try both fixes but no solution ...

**Comment by almic** [comment]
fixed now in d67ea481af1e8528c617dfd404d0c7827b0134b2
